File: Poly/Actor.cs
using System.Drawing;
using Game.Interface;
using OpenTK.Graphics.OpenGL4;
using Game.Math;

namespace Poly;

public class Actor : IActor
{
  private readonly int _buffer = GL.GenBuffer();
  private int _bufferSize;
  private Shader _shader = default!;
  private Vector2 _position;
  private int _index;
  private float _alpha = 1.0f;
  private float _angle;
  private float _size = 1.0f;
  private Color _color;
  public Actor (Shader shader)
  {
    _shader = shader;
  }

  public void UploadData(float[] data)
  {
    GL.BindBuffer(BufferTarget.ArrayBuffer, _buffer);
    GL.BufferData(BufferTarget.ArrayBuffer, data.Length * sizeof(float), data, BufferUsageHint.StaticDraw);
    _bufferSize = data.Length / 2;
  }

  public void Position(Vector2 cord)
  {
    _position = cord;
  }

  public void Rotation(float degrees)
  {
    _angle = degrees;
  }

  public void Transparency(float alpha = 1)
  {
    _alpha = alpha;
  }

  public void Index(int index = 0)
  {
    _index = index;
  }

  public void Scale(float size)
  {
    _size = size;
  }
  
  public void Color(Color color)
  {
    _color = color;
  }
  public void Draw()
  {
    GL.BindBuffer(BufferTarget.ArrayBuffer, _buffer);
    _shader.Active();
    _shader.Position(_position.X, _position.Y);
    _shader.Rotate(_angle);
    _shader.Alpha(_alpha);
    _shader.Index(_index);
    _shader.Size(_size, _size);
    _shader.Color(_color.R/255f, _color.G/255f, _color.B/255f);
    GL.DrawArrays(PrimitiveType.Triangles, 0, _bufferSize);
  }
}

File: BlazorWasm/wwwroot/Rendering/Actor.js
class Actor {
    #buffer
    #bufferSize
    #shader
    #position
    #index
    #alpha
    #angle
    #size
    #color

    constructor(shader) {
        this.#buffer = gl.createBuffer()
        this.#bufferSize = 0
        this.#shader = shader
        this.#position = {x: 0, y: 0}
        this.#index = 0
        this.#alpha = 1.0
        this.#angle = 0
        this.#size = {x: 1.0, y: 1.0}
        this.#color = {r: 0, g: 0, b: 0}
    }

    uploadData(data) {
        var f32data = new Float32Array(data)
        gl.bindBuffer(gl.ARRAY_BUFFER, this.#buffer)
        gl.bufferData(gl.ARRAY_BUFFER, f32data, gl.STATIC_DRAW)
        this.#bufferSize = f32data.length / 2
    }

    position(x, y) {
        this.#position = {x, y}
    }

    rotation(degrees) {
        this.#angle = degrees
    }

    transparency(alpha = 1) {
        this.#alpha = alpha
    }

    index(index = 0) {
        this.#index = index
    }

    scale(x, y) {
        this.#size = {x, y}
    }

    color(r, g, b) {
        this.#color = {r, g, b}
    }

    draw() {
        gl.bindBuffer(gl.ARRAY_BUFFER, this.#buffer)
        this.#shader.active()
        this.#shader.position(this.#position.x, this.#position.y)
        this.#shader.rotate(this.#angle)
        this.#shader.alpha(this.#alpha)
        this.#shader.index(this.#index)
        this.#shader.size(this.#size.x, this.#size.y)
        this.#shader.color(this.#color.r / 255, this.#color.g / 255, this.#color.b / 255)
        gl.drawArrays(gl.TRIANGLES, 0, this.#bufferSize)
    }
}